synapse/util/threepids.py

```diff
@@ -48,3 +51,30 @@ def check_3pid_allowed(hs, medium, address):
         return True
 
     return False
+
+
+def canonicalise_email(address) -> str:
+    """'Canonicalise' email address
+    Case folding of local part of email address and lowercase domain part
+    See MSC2265, https://github.com/matrix-org/matrix-doc/pull/2265
+
+    Args:
+        address (str): email address within that medium (e.g. "wotan@matrix.org")
+    Returns:
+        (str) The canonical form of the email address
+    Raises:
+        SynapseError if the address could not be parsed.
```

Comment:
I don't think it's correct that utility functions like this - which could be used anywhere - should decide what HTTP error code to return. You should raise a ValueError or similar, and turn it into a SynapseError in the rest layer.

Reply:
I am not sure if I should catch error in `auth.py`.
https://github.com/matrix-org/synapse/pull/7021/commits/27866381f18d11c8fb44607ad8ee8c0f34364321